cpp/tests/iterator/iterator_test.cu

```diff
@@ -0,0 +1,393 @@
+    std::cout << "expected <null_iterator> = " << expected_value << std::endl;
+
+    if(1)
+    {  // check host side `IteratorWithNulls`.
+        ColumnData<T> col(hos_array.data(), host_nulls.data(), init);
+
+        column_input_iterator<T, ColumnData<T>> it_hos(col);
+//        this->iterator_test_thrust(expected_value, it_hos, dev_array.size(), false);
+    }
+
+    ColumnData<T> col(dev_array.data().get(), dev_nulls.data().get(), init);
+    column_input_iterator<T, ColumnData<T>> it_dev(col);
+
+    // reduction using thrust
+    this->iterator_test_thrust(expected_value, it_dev, dev_array.size());
+    // reduction using cub
+    this->iterator_test_cub(expected_value, it_dev, dev_array.size());
+
+    std::cout << "test done." << std::endl;
+}
+
+/* tests for square input iterator
+*/
+TYPED_TEST(IteratorTest, null_iterator_square)
+{
+    using T = int32_t;
+    T init = T{0};
+
+    std::vector<T> hos_array({0, 6, 0, -14, 13, 64, -13, -20, 45});
+    thrust::device_vector<T> dev_array(hos_array);
+
+    std::vector<bool> host_bools({1, 1, 0, 1, 1, 1, 0, 1, 1});
+    std::vector<gdf_valid_type> host_nulls;
+    gen_nullbitmap(host_nulls, host_bools);
+    thrust::device_vector<gdf_valid_type> dev_nulls(host_nulls);
+
+    EXPECT_EQ(hos_array.size(), host_bools.size());
+
+    // calculate the expected value by CPU.
+    std::vector<T> replaced_array(hos_array.size());
+    std::transform(hos_array.begin(), hos_array.end(), host_bools.begin(),
+        replaced_array.begin(), [&](T x, bool b) { return (b)? x*x : init; } );
+    T expected_value = std::accumulate(replaced_array.begin(), replaced_array.end(), init);
+    std::cout << "expected <null_iterator> = " << expected_value << std::endl;
+
+    if(1)
+    {
+        ColumnDataSquare<T> col(hos_array.data(), host_nulls.data(), init);
+        column_input_iterator<T, ColumnDataSquare<T>> it_hos(col);
+//        this->iterator_test_thrust(expected_value, it_hos, dev_array.size(), false);
+    }
+
+    ColumnDataSquare<T> col(dev_array.data().get(), dev_nulls.data().get(), init);
+    column_input_iterator<T, ColumnDataSquare<T>> it_dev(col);
+
+    // reduction using thrust
+    this->iterator_test_thrust(expected_value, it_dev, dev_array.size());
+    // reduction using cub
+    this->iterator_test_cub(expected_value, it_dev, dev_array.size());
+
+    std::cout << "test done." << std::endl;
+}
+
+
+/*
+    tests for group_by iterator
+    this was used by old implementation of group_by.
+
+    This won't be used with the newer implementation
+     (a.k.a. Single pass, distributive groupby https://github.com/rapidsai/cudf/pull/1478)
+    distributive groupby uses atomic operation to accumulate.
+
+    For group_by.cumsum() (scan base group_by) may not be single pass scan.
+    There is a possiblity that this process may be used for group_by.cumsum().
+*/
+TYPED_TEST(IteratorTest, group_by_iterator)
```

Comment:
I don't understand how this test relates to `group_by`. This just looks to be doing a reduction, same as the other tests.

Reply:
This does a reduction for the selected indices.
Let's say, .group_by('A') where 'A' = [1, 1, 2,  1, 2, 1,  2, 2, 2]
The indices for `1` => [0, 1, 3, 5], `2` => [2, 4, 6, 7, 8]
We actually do 2 reductions for the 2 indices instead of distributed single reduction.

I know group_by.sum() never uses reduction for the selected indices.
However, how do we implement group_by.cumsum() (scan base group_by)?
I feel there is a possibility that this process may be used for group_by.cumsum().
Of course, this may not be used if there is better way.

Reply:
Oh! I see, I missed what was going on here.

You're passing an array of `[0,1,3,5]` as the base iterator for `column_input_iterator` which will only perform the reduction on those indices! Makes sense now. 

That said, I think this would end up being an expensive way to implement a groupby since you'd be doing a separate kernel call for each unique group. Furthermore, you'd have to first construct a list of indices where values belonging to each group reside. 

When I've thought about implementing a `groupby.cumsum()`, I'd thought to first sort the data, and then use a `thrust::exclusive_scan_by_key`.

Reply:
> That said, I think this would end up being an expensive way to implement a groupby since you'd be doing a separate kernel call for each unique group. Furthermore, you'd have to first construct a list of indices where values belonging to each group reside.
> 
> When I've thought about implementing a `groupby.cumsum()`, I'd thought to first sort the data, and then use a `thrust::exclusive_scan_by_key`.

`thrust::exclusive_scan_by_key` looks good. I missed it.
Okay, I will leave this test exist since it test the feasibility of indexed access. But I will add a comment "this is a feasibility test, this will be slow and may not be used in cudf real case."